### PR TITLE
ENH: Added serialization for IncrementalBasicStatistics

### PR DESCRIPTION
**Pull request by olegkkruglov**
## Description

* Added `__getstate__` method to `IncrementalBasicStatistics`
* Finalization is called with serialization during `__getstate__` method
* `_need_to_finalize` flag is added to avoid unnecessary call of finalization backend.
* Added test for checking if it works well

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: onedal/basic_statistics/incremental_basic_statistics.py [hunk 1]
```diff
@@ -70,10 +70,27 @@ def __init__(self, result_options="all"):
...
+        self._need_to_finalize = False
         self._partial_result = self._get_backend(
             "basic_statistics", None, "partial_compute_result"
         )
 
+    def __getstate__(self):
+        """
+        Converts estimator's data to serializable format.
+        All tables contained in partial result are converted to np.arrays.
+        Notes
```
Comment (david-cortes-intel) [inline onedal/basic_statistics/incremental_basic_statistics.py:82]: I think the format requirement for rendering this docstring would be to have the title as "Note" and with a newline above it.

I'd also say that this could be better left as a comment than as a docstring, since `__getstate__` is a built-in method in python.

File: onedal/basic_statistics/tests/test_incremental_basic_statistics.py [hunk 2]
```diff
@@ -189,3 +189,60 @@ def test_all_option_on_random_data(
...
+
+    assert incbs._need_to_finalize == True
+    assert incbs_loaded._need_to_finalize == True
+
+    # Check that estmator can be serialized after partial_fit call.
+    dump = pickle.dumps(incbs_loaded)
+    incbs_loaded = pickle.loads(dump)
+    assert incbs._need_to_finalize == True
+    # Finalize is called during serialization to make sure partial results are finalized correctly.
+    assert incbs_loaded._need_to_finalize == False
```
Comment (david-cortes-intel) [inline onedal/basic_statistics/tests/test_incremental_basic_statistics.py:222]: Here it could compare the saved statistics in `incbs` and `incbs_loaded` after this line.

File: sklearnex/basic_statistics/tests/test_incremental_basic_statistics.py [hunk 3]
```diff
@@ -397,3 +397,53 @@ def test_warning():
...
+    seed = 77
+    gen = np.random.default_rng(seed)
+    X = gen.uniform(low=-0.3, high=+0.7, size=(10, 10))
+    X = X.astype(dtype)
+    X_split = np.array_split(X, 2)
+    X_split_df = _convert_to_dataframe(X_split[0], sycl_queue=queue, target_df=dataframe)
+    incbs.partial_fit(X_split_df)
+    incbs_loaded.partial_fit(X_split_df)
+
+    # Check that estmator can be serialized after partial_fit call.
```
Comment (david-cortes-intel) [inline sklearnex/basic_statistics/tests/test_incremental_basic_statistics.py:423]:
```suggestion
    # Check that estimator can be serialized after partial_fit call.
```

File: onedal/basic_statistics/tests/test_incremental_basic_statistics.py [hunk 4]
```diff
@@ -189,3 +189,60 @@ def test_all_option_on_random_data(
             gtr = function(data)
         tol = fp32tol if res.dtype == np.float32 else fp64tol
         assert_allclose(gtr, res, atol=tol)
+
+
+@pytest.mark.parametrize("queue", get_queues())
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+def test_incremental_estimator_pickle(queue, dtype):
```
Comment (david-cortes-intel) [inline onedal/basic_statistics/tests/test_incremental_basic_statistics.py:196]: Since this tests serialization, it should also test that changes in regular attributes like `batch_size` or `n_jobs` remain after deserializing.

File: onedal/basic_statistics/incremental_basic_statistics.py (hunk 1 above)
Comment (olegkkruglov) [reply onedal/basic_statistics/incremental_basic_statistics.py:82]: I think it is better to keep as docstring since I have been requested to add info from the Note to documentation. Probably the best way to do it is usage of `automethod`. Let me know if you have any objections against it.

Comment (david-cortes-intel) [reply onedal/basic_statistics/incremental_basic_statistics.py:82]: Why would that have been requested in the first place? It is rather uncommon to have docstrings for built-ins like these, and methods with two underscores shouldn't show up in autoclass docs.

For example, scikit-learn for the most part doesn't put docstrings on their `__getstate__`/`__setstate__` methods:
https://github.com/search?q=repo%3Ascikit-learn%2Fscikit-learn%20__getstate__&type=code

Comment (david-cortes-intel) [reply onedal/basic_statistics/incremental_basic_statistics.py:82]: In that case, you might want to add that info as a "note" section in the class docstring, because it otherwise wouldn't reach the final documentation.

This is what the docs show for me after building them from your branch - you'll see that autoclass doesn't list double-underscore methods:
image

File: sklearnex/basic_statistics/incremental_basic_statistics.py [hunk 5]
```diff
@@ -103,6 +103,12 @@ class IncrementalBasicStatistics(IntelEstimator, BaseEstimator):
     ----
     Attribute exists only if corresponding result option has been provided.
 
+    Note
+    ----
+    Since finalize_fit can't be dispatched without directly provided queue
```
Comment (david-cortes-intel) [inline sklearnex/basic_statistics/incremental_basic_statistics.py:108]: Thanks for adding this note. But it looks out of context without having seen the PR before - e.g. it doesn't quite convey the idea that something happens during calls to pickle, and that this something is for device data or sycl queues.

Perhaps you could rephrase it along the lines of "serializing instances of this class fitted to sycl data will trigger a forced finalization of calculations" or something similar, otherwise it's not very clear for a final user what all of this means.

File: onedal/basic_statistics/basic_statistics.cpp [hunk 6]
```diff
@@ -204,7 +207,33 @@ void init_partial_compute_result(py::module_& m) {
...
+                if (py::cast<int>(t[1].attr("size")) != 0) res.set_partial_min(convert_to_table(t[1].ptr()));
+                if (py::cast<int>(t[2].attr("size")) != 0) res.set_partial_max(convert_to_table(t[2].ptr()));
+                if (py::cast<int>(t[2].attr("size")) != 0) res.set_partial_sum(convert_to_table(t[3].ptr()));
+                if (py::cast<int>(t[2].attr("size")) != 0) res.set_partial_sum_squares(convert_to_table(t[4].ptr()));
+                if (py::cast<int>(t[2].attr("size")) != 0) res.set_partial_sum_squares_centered(convert_to_table(t[5].ptr()));
+                
+                return res;
+            }
+        ));
+
```
Comment (icfaust) [inline onedal/basic_statistics/basic_statistics.cpp:236]: Please address this codefactor issue (remove the blank line).

**Review by icfaust — COMMENTED** [review]
small changes, when @david-cortes-intel 's issues are addressed this can be merged.

**Review by david-cortes-intel — APPROVED** [review]

**Comment by david-cortes-intel** [comment]
LGTM, provided the test failures are fixed.
